server/src/main/java/org/opensearch/index/shard/IndexShard.java

```diff
@@ -4661,36 +4657,15 @@ public void syncSegmentsFromRemoteSegmentStore(boolean overrideLocal, boolean re
                         indexInput,
                         remoteSegmentMetadata.getGeneration()
                     );
-                    // Replicas never need a local commit
-                    if (shouldCommit) {
-                        if (this.shardRouting.primary()) {
-                            long processedLocalCheckpoint = Long.parseLong(infosSnapshot.getUserData().get(LOCAL_CHECKPOINT_KEY));
-                            // Following code block makes sure to use SegmentInfosSnapshot in the remote store if generation differs
-                            // with local filesystem. If local filesystem already has segments_N+2 and infosSnapshot has generation N,
-                            // after commit, there would be 2 files that would be created segments_N+1 and segments_N+2. With the
-                            // policy of preserving only the latest commit, we will delete segments_N+1 which in fact is the part of the
-                            // latest commit.
-                            Optional<String> localMaxSegmentInfos = localSegmentFiles.stream()
-                                .filter(file -> file.startsWith(IndexFileNames.SEGMENTS))
-                                .max(Comparator.comparingLong(SegmentInfos::generationFromSegmentsFileName));
-                            if (localMaxSegmentInfos.isPresent()
-                                && infosSnapshot.getGeneration() < SegmentInfos.generationFromSegmentsFileName(localMaxSegmentInfos.get())
-                                    - 1) {
-                                // If remote translog is not enabled, local translog will be created with different UUID.
-                                // This fails in Store.trimUnsafeCommits() as translog UUID of checkpoint and SegmentInfos needs
-                                // to be same. Following code block make sure to have the same UUID.
-                                if (indexSettings.isRemoteTranslogStoreEnabled() == false) {
-                                    SegmentInfos localSegmentInfos = store.readLastCommittedSegmentsInfo();
-                                    Map<String, String> userData = new HashMap<>(infosSnapshot.getUserData());
-                                    userData.put(TRANSLOG_UUID_KEY, localSegmentInfos.userData.get(TRANSLOG_UUID_KEY));
-                                    infosSnapshot.setUserData(userData, false);
-                                }
-                                storeDirectory.deleteFile(localMaxSegmentInfos.get());
+                    long processedLocalCheckpoint = Long.parseLong(infosSnapshot.getUserData().get(LOCAL_CHECKPOINT_KEY));
+                    store.commitSegmentInfos(infosSnapshot, processedLocalCheckpoint, processedLocalCheckpoint);
+                    Collection<String> currentInfoFiles = infosSnapshot.files(true);
+                    if (recoveryState.getStage() != RecoveryState.Stage.DONE) {
```

Comment:
Can we use IndexShardState here instead?

Reply:
Yes, we can use it and is better to use here since, we need to perform this cleanup only for non-started shards. Updated to `if (state() != IndexShardState.STARTED) {`